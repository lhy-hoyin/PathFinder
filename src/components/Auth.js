import { useState } from "react";
import { supabase } from "../supabaseClient";

export default function Auth() {
  const [isLoading, setLoading] = useState(false);
  const [email, setEmail] = useState("");

  const handleLogin = async e => {
    e.preventDefault();

    try {
      setLoading(true);

      const { error } = await supabase.auth.signIn({ email });
      if (error) throw error;
      alert("Check your email for the login link!");
    } catch (error) {
      alert(error.error_description || error.message);
    } finally {
      setLoading(false);
    }
  };

  return (
    <auth>
      <h2>Let's Get Started</h2>
      <form onSubmit={handleLogin}>
        <label htmlFor="email">Email</label>
        <input
          id="email"
          className="inputField"
          type="email"
          placeholder="Your email"
          value={email}
          onChange={e => setEmail(e.target.value)}
        />

        <button className="button block" aria-live="polite">
          Sign Up
        </button>

        <button className="button block" aria-live="polite">
          Login using email
        </button>
      </form>

      {isLoading ? <p>Signing up, please wait</p> : ""}
      <p>{email}</p>
    </auth>
  );
}
