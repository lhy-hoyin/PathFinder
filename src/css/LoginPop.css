@import "styles.css";




.login-popup{
    text-align: center;
    height: 150px;
    width: 100%;
    position: relative;
    font-size: 18px;
    display: flex;
    border: 10cm; 
}

.inside-login-popup{
    width: 100%;
    color: #686565;
    font-size: 20px;
    font-family: monospace;
    
    

}