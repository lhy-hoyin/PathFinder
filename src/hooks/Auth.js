import { useState, useEffect, useContext, createContext } from "react";
import { supabase } from "../supabaseClient";

const authContext = createContext();


export function ProvideAuth({ children }) {
    const auth = useProvideAuth();
    return <authContext.Provider value={auth}>{children}</authContext.Provider>;
}

export const Auth = () => {
    return useContext(authContext);
};

function useProvideAuth() {
    const [session, setSession] = useState(null);
    const [profileInfoReady, setProfileInfoReady] = useState(false);

    // User-releated info
    const [email, setEmail] = useState(null);
    const [firstName, setFirstName] = useState(null);
    const [lastName, setLastName] = useState(null);
    const [cohort, setCohort] = useState(null);
    const [isReady, setIsReady] = useState(null);
    const [isLocked, setIsLocked] = useState(null);

    useEffect(() => {
        setSession(supabase.auth.session());
        supabase.auth.onAuthStateChange((_event, session) => {setSession();})
    }, []);

    useEffect(() => {
        getProfile();
    }, [session]);

    const getProfile = async () => {
        try {
            const user = supabase.auth.user();

            if (user == null)
                return

            setEmail(user.email);

            let { data, error, status } = await supabase
                .from('profiles')
                .select('*')
                .eq('id', user.id)
                .single()

            if (error && status !== 406)
                throw error

            if (data) {
                setFirstName(data.FirstName);
                setLastName(data.LastName);
                setCohort(data.Cohort);
                setIsReady(data.isReady);
                setIsLocked(data.isLocked);

                setProfileInfoReady(true);
            }
        } catch (error) {
            console.error(error.message);
        }
    };

    const updateProfileBasic = (fName, lName) => async e => {
        e.preventDefault();

        try {
            const user = supabase.auth.user()

            if (user == null)
                return

            // Pacakage data properly
            const updates = {
                id: user.id,
                FirstName: fName,
                LastName: lName,
                isReady: true,
                updated_at: new Date(),
            }

            let { error } = await supabase.from('profiles').upsert(updates, {
                returning: 'minimal', // Don't return the value after inserting
            })

            if (error)
                throw error
            else
                console.log("Profile Updated successfully")

        } catch (error) {
            console.error(error.message);
        }
    };

    const updateProfileAcad = (cohort) => async e => {
        e.preventDefault();

        try {
            const user = supabase.auth.user()

            if (user == null)
                return

            // Pacakage data properly
            const updates = {
                id: user.id,
                Cohort: cohort,
                updated_at: new Date(),
            }

            let { error } = await supabase.from('profiles').upsert(updates, {
                returning: 'minimal', // Don't return the value after inserting
            })

            if (error)
                throw error
            else
                console.log("Profile Updated successfully")

        } catch (error) {
            console.error(error.message);
        }
    };

    const signup = (email, password1, password2, setMessage) => async e => {
        e.preventDefault();

        // Verify that password is matching
        if (password1 != password2) {
            setMessage("Password does not match");
            console.log("User input mis-matched password");
            return;
        }

        var password = password1;

        if (password.length < 6) {
            setMessage("Password too short (min 6 sharacters)");
            console.log("User password too short");
            return;
        }

        //TODO: hex password here for security
        //note: login also need to hex, when this is implemented

        try {
            setMessage("Signing up ... please be patient...");
            const { error } = await supabase.auth.signUp({ email, password });
            if (error) throw error;
            setMessage("Check your email for the login link!");
            console.log("Verification email sent");
        } catch (error) {
            console.error(error.error_description || error.message);
            setMessage(error.error_description || error.message);
        }
    };

    const login = (email, password) => async e => {
        e.preventDefault();

        //TODO: hex password here for security
        //note: singup also need to hex, when this is implemented

        try {
          const { error } = await supabase.auth.signIn({ email, password });
          if (error) throw error;
            console.log("User logged in");
        } catch (error) {
            alert(error.error_description || error.message);
            console.error(error.error_description || error.message);
        }
    };

    const logout = async e => {
        e.preventDefault();

        try {
            const { error } = await supabase.auth.signOut();
            if (error) throw error;
            console.log("User logged out");
            alert("User logged out");
            
            window.location = window.location.origin.toString();
        } catch (error) {
            console.error(error.error_description || error.message);
        }
    };  

    const sendPasswordReset = (email, setMessage) => async e => {

        e.preventDefault();
        try {
            setMessage("Sending recovery link....please wait")

            const { data, error } = await supabase.auth.api.resetPasswordForEmail(email, {
                redirectTo: `${window.location.origin}/reset-password`,
              });
            if (error) throw error;

            alert("Recovery link has been sent to your email");
            setMessage("Link has been sent!");
        } catch (error) {
            setMessage("Email does not exist");
            console.error(error.error_description);
        }
    };

    const resettingPassword = (password1, password2, setMessage) => async e => {
        e.preventDefault();

        if (password1 != password2) {
            setMessage("Password does not match");
            console.log("User input mis-matched password");
            return;
        }

        var password = password1

        try {
            const {error } = await supabase.auth.update({password});
            if (error) throw error;
            alert("password changed ");
        } catch (error) {
            alert(error.error_description);
            console.error(error.error_description);
        }
    };

    return {
        // Account-releated functions
        signup,
        login,
        logout,
        sendPasswordReset,
        resettingPassword ,
        updateProfileBasic,
        updateProfileAcad,

        // Status of profile info
        profileInfoReady,

        // User-releated info
        email,
        firstName,
        lastName,
        cohort,
        isReady,
        isLocked,
    };
}