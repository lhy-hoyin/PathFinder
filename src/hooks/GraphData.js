import { useState, useContext, createContext } from "react";
import cloneDeep from 'lodash/cloneDeep';
import Color from "color";

import { supabase } from "../supabaseClient";
import { ModuleStateColor } from "../constants"

const graphContext = createContext();

export function ProvideGraphData({ children }) {
    const graphData = useProvideGraphData();
    return <graphContext.Provider value={graphData}>{children}</graphContext.Provider>;
}

export const graphData = () => {
    return useContext(graphContext);
};

function useProvideGraphData() {

    const [modules, setModules] = useState([]);
    const [preq, setPreq] = useState([]);

    const colouring = (selectedColor) => {
        const color ={
           border: Color(selectedColor).darken(0.2).hex(),
           background : selectedColor,
           highlight: {
             border: Color(selectedColor).darken(0.3).hex(),
             background: Color(selectedColor).darken(0.2).hex()
            },
            hover: {
                border: Color(selectedColor).darken(0.3).hex(),
                background: Color(selectedColor).darken(0.2).hex()
            }
        }
         
         return color;
    }

    const orNodesLabel = (orArray) => {
        let label = orArray[0].toString();
        for (var x = 1; x < orArray.length; x++) {
          label = label + " or " + orArray[x]
        }
        return label
    }

    const nodeExist = (checkMod, arr) => {
        let modExist = [];
        let count = 0;
    
        for (var x = 0; x < checkMod.length; x++) {
          if (arr.includes(checkMod[x])) {
            modExist[count] = checkMod[x];
            count++;
          }
        }
        return modExist;
      };

    const updateStatus = (allMods, selectedModule) => {
        let index = allMods.findIndex((x) => x.id === selectedModule);
    
        if (allMods[index].status === true) {
          allMods[index].color = colouring(ModuleStateColor.Completed);
          return allMods;
        }
        //Starting modules that have no prequities
        if (allMods[index].pre.length === 0) {
          if (allMods[index].status === false) {
            allMods[index].color = colouring(ModuleStateColor.Available);
          } else {
            allMods[index].color = colouring(ModuleStateColor.Completed);
          }
        } else {
          //Modules with preq
          for (var count = 0; count < allMods[index].pre.length; count++) {
            
            //For or nodes
            if (allMods[index].pre[count][0].length > 1) {
              for ( var count1 = 0; count1 < allMods[index].pre[count].length; count1++) {
                let temp = allMods.findIndex( (x) => x.id === allMods[index].pre[count][count1]);
                if (allMods[temp].status === true) {
                  allMods[index].color = colouring(ModuleStateColor.Available);
                  return allMods;
                }
              }
              allMods[index].color = colouring(ModuleStateColor.Locked);
              return allMods;
            } else {
              let temp = allMods.findIndex((x) => x.id === allMods[index].pre[count]);

              if (allMods[temp].status === false) {
                allMods[index].color = colouring(ModuleStateColor.Locked);
                return allMods;
              }
            }
          }
    
          allMods[index].color = colouring(ModuleStateColor.Available);
        }
    
        return allMods;
      };

    const getData = (modsArr) => async e => {
        e.preventDefault();

        try {
            let { data, error } = await supabase
                .from("modules")
                .select("*")
                .filter('code', 'in', `(${modsArr})`)

            if (data == null)
                throw ("no data from database")

            let temp = data.slice(0);
            const count = temp.length

            let mod = [];
            let orNodes = [];
            let orCount = 0;

            let edges = [];
            let edgeCount = 0;

            for (var node = 0; node < count; node++) {

                mod[node] = {
                    id: temp[node].code,
                    label: temp[node].code,
                    color: colouring(ModuleStateColor.Completed),
                    info: [temp[node].name, temp[node].acad_year, temp[node].credit, temp[node].description],
                    status: false,
                    pre: []
                };

                if (temp[node].pre_req) {

                    for (var req = 0; req < temp[node].pre_req.length; req++) {
                        
                        let modWithOr = temp[node].pre_req[req].toString().split(",");
                        modWithOr = nodeExist(modWithOr, modsArr)

                        if (modWithOr.length > 1) {

                            orNodes[orCount] = {
                                id:  orNodesLabel(modWithOr),
                                label: orNodesLabel(modWithOr),
                                shape: "ellipse",
                                color: colouring(ModuleStateColor.Completed),
                                info: ["", "", "", ""],
                                status: false,
                                pre: [modWithOr]
                            }
                            orCount++;

                            mod[node].pre.push(orNodesLabel(modWithOr));
                            edges[edgeCount] = { from: temp[node].code, to: orNodesLabel(modWithOr) }
                            edgeCount++;

                            for(var orReq = 0; orReq < modWithOr.length; orReq++) {
                                edges[edgeCount] = { from: orNodesLabel(modWithOr) , to: modWithOr[orReq] }
                                edgeCount++;
                            }

                        } else{
                            if (modWithOr.toString() !== "") {
                                mod[node].pre.push(modWithOr.toString());
                                edges[edgeCount] = { from: temp[node].code, to: modWithOr.toString() };
                                edgeCount++;
                            }
                            
                        }                       
                    }
                }
            }

            mod = mod.concat(orNodes)

            for (var x = 0; x < mod.length; x++) {
                updateStatus(mod, mod[x].id);
            }

            setPreq(edges); // need to set edges first
            setModules(mod);

        } catch (error) {
            alert(error.message);
        }
    };

    const getCourses = async () => {
        try {
            let { data, error } = await supabase
                .from("courses")
                .select("course_name")

            if (data == null)
                throw error 

            return data.map(row => row.course_name)

        } catch (error) {
            console.error(error.message);
        }
    }

    const updateGraph = (nodes) => {
        let index = modules.findIndex((x) => x.id === nodes.toString());
        if (modules[index].color.background === ModuleStateColor.Locked) {
          alert("Module is locked. Please clear the prequities first.");
        } else {
            const modulesCopy = cloneDeep(modules);
            modulesCopy[index].status = !modulesCopy[index].status;
            for (var x = 0; x < modulesCopy.length; x++) {
                updateStatus(modulesCopy, modulesCopy[x].id);
            }
            setModules(modulesCopy);
          //  isUpdating(!updating)
        }
      };

    return {
        getData,
        getCourses,
        updateGraph,

        modules,
        preq
    };
}