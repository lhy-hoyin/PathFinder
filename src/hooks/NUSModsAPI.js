const BASE_URL = "https://api.nusmods.com/v2/";

const generateQueryUrl = (moduleCode, acadYear) => {
	if (moduleCode === undefined) {
		console.error("getMdule: no moduleCode to search")
		return undefined
	}

	if (acadYear === undefined) {
		const currentYear = new Date().getFullYear()
		acadYear = currentYear.toString() + "/" + (currentYear + 1).toString()
	}

	return BASE_URL
		+ acadYear.replace("/", "-").toUpperCase().replace("AY", "")
		+ "/modules/"
		+ moduleCode.toUpperCase()
		+ ".json"
}

export const moduleExist = async (moduleCode, acadYear) => {
	const queryUrl = generateQueryUrl(moduleCode, acadYear) 
	return fetch(queryUrl)
		.then(res => { return res.ok })
};

export const getModule = async (moduleCode, acadYear) => {
	const queryUrl = generateQueryUrl(moduleCode, acadYear) 
	return fetch(queryUrl)
		.then(res => res.json())
		.then(
			(result) => { return { data: result } },
			(err) => { return { error: err } }
		)
};