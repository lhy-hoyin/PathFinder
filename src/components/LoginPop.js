import { useState } from "react";
import {
    Input, InputGroup, InputRightElement,
    Text, Button, useBoolean
} from '@chakra-ui/react';

import { Auth } from "../hooks/Auth";

import "../css/styles.css";

export default function LoginPop() {

    const { login } = Auth();

    const [email, setEmail] = useState("");
    const [password, setPassword] = useState("");

    const [showPassword, setShowPassword] = useBoolean();

    return (
        <div >
            <form onSubmit={login(email, password)}>

                <Text> Login Details</Text>

                <Input
                    id="email"
                    data-testid="email_input"
                    type="email"
                    placeholder="Email"
                    autoComplete="email"
                    required
                    margin={1}
                    value={email}
                    onChange={(e) => setEmail(e.target.value)} />

                <InputGroup size='md' margin={1}>
                    <Input
                        id="password"
                        data-testid="password-input"
                        type={showPassword ? 'text' : 'password'}
                        placeholder="Enter Password"
                        autoComplete="password"
                        paddingRight='4.5rem'
                        required
                        value={password}
                        onChange={(e) => setPassword(e.target.value)} />
                    <InputRightElement width='4.5rem'>
                        <Button h='1.75rem' size='sm' onClick={setShowPassword.toggle}>
                            {showPassword ? 'Hide' : 'Show'}
                        </Button>
                    </InputRightElement>
                </InputGroup>

                <Button
                    data-testid="login-btn"
                    type="submit"
                    colorScheme='blue'
                    margin={1}>
                    Log In
                </Button>

            </form>

            <a className='clickable' href="/forget-password">Forget Password?</a>

        </div>
    );
}