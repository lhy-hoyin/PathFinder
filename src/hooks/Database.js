import { supabase } from "../supabaseClient";
import { pullModule, formatPreReq } from "../hooks/NUSModsAPI";

export const getCourseNames = async () => {
    try {
        let { data, error } = await supabase
            .from("courses")
            .select("course_name")

        if (data == null)
            throw error

        return data.map(row => row.course_name)

    } catch (error) {
        console.error(error.message)
    }
};

export const getModuleId = async (modCode, year) => {
    try {
        let { data, status, error } = await supabase
            .from("modules")
            .select("id")
            .eq('acad_year', year)
            .eq('code', modCode.toUpperCase())
            .single()

        if (error && status == 406) {
            /*
             * Error 406: multiple (or no) rows returned
             * No row fround is expected, do nothing here is suppress it
             */
        } else if (data == null)
            throw error

        return data

    } catch (error) {
        console.error(error.message)
        return null
    }
};

export const getModInfo = async (modId) => {
    try {
        if (modId == undefined || modId == null)
            throw { message: "getModInfo: modId is undefined" }

        let { data, error } = await supabase
            .from("modules")
            .select("*")
            .eq('id', modId)
            .single()

        if (data == null)
            throw error

        return data

    } catch (error) {
        console.error(error.message)
    }
};

export const upsertModule = async (moduleCode, acadYear) => {
    try {
        // Retrieve data from API
        const modData = (await pullModule(moduleCode, acadYear)).data

        // Package data properly
        const updates = {
            code: modData.moduleCode,
            name: modData.title,
            description: modData.description,
            acad_year: modData.acadYear,
            credit: modData.moduleCredit,
            preclusion: modData.preclusion,
            pre_req: formatPreReq(modData.prereqTree),
            updated_at: new Date(),
        }

        // if table has existing entry, add the id to update the correct entry
        const existingRowId = await getModuleId(moduleCode, acadYear)
        if (existingRowId)
            updates.id = existingRowId.id

        let { data, error } = await supabase.from('modules').upsert(updates)

        if (error)
            throw error

        console.log("Database Updated:", moduleCode, acadYear)
        return data[0].id

    } catch (error) {
        console.error(error.message);
    }
};

export const getUserAcademic = async (userId) => {
    try {
        let { data, error } = await supabase
            .from("academic")
            .select("*")
            .eq('user_id', userId)

        if (data == null)
            throw error

        return data

    } catch (error) {
        console.error(error.message)
    }
};

export var UserAcademicRecord = {
    insert: async (userId, newModId) => {
        try {
            const { error } = await supabase
                .from('academic')
                .insert([{
                    user_id: userId,
                    module: newModId,
                }])

            if (error)
                throw error
            else
                return { status: 'success' }

        } catch (error) {
            console.error(error.error_description || error.message)
            return {
                status: 'error',
                title: "Oops!",
                description: error.error_description || error.message
            }
        }
    },

    update: async (recordId, isCompleted) => {

        try {
            const { error } = await supabase
                .from('academic')
                .update({ completed: isCompleted })
                .eq('id', recordId)

            if (error)
                throw error
        } catch (error) {
            console.error(error.error_description || error.message);
        }
    },

    delete: async (recordId) => {
        try {
            const { status, error } = await supabase
                .from('academic')
                .delete()
                .eq('id', recordId)
                .single()

            return { status, error }
        } catch (error) {
            console.error(error.message)
            return { status, error }
        }
    },
}