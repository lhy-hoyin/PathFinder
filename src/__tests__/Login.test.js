import { cleanup, fireEvent, render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { ProvideAuth } from "../hooks/Auth";
import LoginPop from "../components/LoginPop";

afterEach(cleanup);


it('renders without crashing', () => {
    const tree = renderer.create(
        <ProvideAuth>
            <LoginPop />
        </ProvideAuth>
    ).toJSON();
    expect(tree).toMatchSnapshot();
});


it("login by normal user", () => {
    const loginMock = jest.fn();

    render(<ProvideAuth><LoginPop /></ProvideAuth>)

    const email = screen.getByTestId("email_input")
    const password= screen.getByTestId("password-input")
    const submit = screen.getByTestId("login-btn")

    fireEvent.change(email, { target: { value: 'vrv84363@xcoxc.com'} })
    fireEvent.change(password, { target: { value: 'password' } })
    fireEvent.click(submit)

    expect(loginMock).toBeCalled()
})