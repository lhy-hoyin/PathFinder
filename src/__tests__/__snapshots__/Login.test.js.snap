// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <p
      className="chakra-text css-0"
    >
       Login Details
    </p>
    <input
      aria-required={true}
      autoComplete="email"
      className="chakra-input css-13d3w74"
      data-testid="email_input"
      id="email"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder="Email"
      required={true}
      type="email"
      value=""
    />
    <div
      className="chakra-input__group css-s4jois"
    >
      <input
        aria-required={true}
        autoComplete="password"
        className="chakra-input css-1n0hj84"
        data-testid="password-input"
        id="password"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Enter Password"
        required={true}
        type="password"
        value=""
      />
      <div
        className="chakra-input__right-element css-1x84ua"
        size="md"
      >
        <button
          className="chakra-button css-xsdlqf"
          onClick={[Function]}
          type="button"
        >
          Show
        </button>
      </div>
    </div>
    <button
      className="chakra-button css-1nrxiz"
      data-testid="login-btn"
      type="submit"
    >
      Log In
    </button>
  </form>
  <a
    className="clickable"
    href="/forget-password"
  >
    Forget Password?
  </a>
</div>
`;
